suse2021-ns/fo/xref.xsl

```diff
@@ -159,30 +159,40 @@
 </xsl:template>
 
 <xsl:template match="d:xref" name="xref">
-  <xsl:variable name="targets" select="key('id',@linkend)"/>
-  <xsl:variable name="target" select="$targets[1]"/>
-  <xsl:variable name="refelem" select="local-name($target)"/>
-  <xsl:variable name="target.book" select="($target/ancestor-or-self::d:article|$target/ancestor-or-self::d:book)[1]"/>
-  <xsl:variable name="this.book" select="(ancestor-or-self::d:article|ancestor-or-self::d:book)[1]"/>
+    <xsl:variable name="context" select="."/>
+    <xsl:variable name="target" select="key('id', @linkend)[1]"/>
+    <xsl:variable name="xref.in.samebook">
+      <xsl:call-template name="is.xref.in.samebook">
+        <xsl:with-param name="target" select="$target"/>
+      </xsl:call-template>
+    </xsl:variable>
+
   <xsl:variable name="lang" select="(ancestor-or-self::*/@lang|ancestor-or-self::*/@xml:lang)[1]"/>
-  <xsl:variable name="xref.in.samebook">
-    <xsl:call-template name="is.xref.in.samebook">
-      <xsl:with-param name="target" select="$target"/>
-    </xsl:call-template>
-  </xsl:variable>
 
   <xsl:call-template name="check.id.unique">
     <xsl:with-param name="linkend" select="@linkend"/>
   </xsl:call-template>
 
   <xsl:choose>
-    <!-- If we point to another book AND we have an xrefstyle present,
-         output the same text content as the original xref template but
-         remove the link.
-    -->
+    <xsl:when test="$rootid = ''">
+        <xsl:apply-imports/>
```

Comment:
>  I think it would be trivial to handle this case within `xref.in.samebook` rather than duplicate it for each output format.

I would prefer to have this template to deal with just *one task*: test if the xref is inside or outside of a specific element.

We could create another template that does these trivial cases, right? (Although to be honest, I'm not entirely sure how could this look like.)

Reply:
* It would still be dealing with just one task -- except it could do that task with or without $rootid being defined.
* Simplifying/reducing the duplicated logic in the HTML/FO stylesheets seems a good goal to me irregardless.
* Essentially, you'd just need the idea from https://github.com/openSUSE/suse-xsl/pull/436#discussion_r698049323 (with `[last()]` not `[1]`, as you correctly noted though)